lib/puppet/etc.rb

```diff
@@ -96,10 +113,9 @@ def convert_field_values_to_utf8!(struct)
       return nil if struct.nil?
       struct.each_with_index do |value, index|
         if value.is_a?(String)
-            converted = Puppet::Util::CharacterEncoding.convert_to_utf_8!(value)
-            struct[index] = converted if !converted.nil?
+          Puppet::Util::CharacterEncoding.override_encoding_to_utf_8!(value) if value.encoding != Encoding::UTF_8
```

Comment:
@joshcooper @Iristyle This change is the crux of our discussion - This now basically says we are explicitly overriding the encoding of the string returned by Etc to UTF-8 (not converting it) if it is a) not already UTF-8 and b) would be valid UTF-8. We are starting from the assumption that it is incorrectly labeled with whatever encoding it has (ie ISO-8859-1) because we have reason to believe it is actually UTF-8 because Puppet set (or will set) the value to UTF-8.

Just to reiterate for PR posterity why I believe this is the correct behavior: Consider these two scenarios. In the first, we default to assuming the encoding is correct from Etc and do `encode!` and in the second we assume the encoding is incorrect and override it with `force_encoding`.

* Puppet agent is creating a new user with managed comment property. The host is running in ISO-8859-1 encoding. When puppet agent get's its `should` value via the catalog, the should value is UTF-8 and its bytes are written to disk as-is with no re-encoding/modification (we don't use Etc for writing, only reading). When Puppet runs next time, Ruby::Etc will read in these bytes and label them as ISO-8859-1, which will likely be valid in that encoding. At least, our commonly used `mixed_utf8` string in tests is "valid" ISO-8859-1. When we pipe these values through Puppet::Etc, they'll now be `encode!`d as UTF-8 - changing the byte representation. So our original UTF-8 has now been modified to something else. When Puppet compares this `is` value to the catalog `should` value, they won't be the same... so Puppet will consider the resource out of sync and write the `should` value to disk again - and will every time it runs.

```
[1] pry(main)> mixed_utf_8 = "A\u06FF\u16A0\u{2070E}"
=> "Aۿᚠ𠜎"
[2] pry(main)> mixed_utf_8.bytes
=> [65, 219, 191, 225, 154, 160, 240, 160, 156, 142]
[3] pry(main)> mixed_utf_8.force_encoding(Encoding::ISO_8859_1)
=> "A\xDB\xBF\xE1\x9A\xA0\xF0\xA0\x9C\x8E"
[4] pry(main)> mixed_utf_8.valid_encoding?
=> true
[5] pry(main)> mixed_utf_8.encode!(Encoding::UTF_8)
=> "AÛ¿á\u009A ð \u009C\u008E"
[6] pry(main)> mixed_utf_8.bytes
=> [65, 195, 155, 194, 191, 195, 161, 194, 154, 194, 160, 195, 176, 194, 160, 194, 156, 194, 142]
[7] pry(main)> orig_mixed_utf_8 = "A\u06FF\u16A0\u{2070E}"
=> "Aۿᚠ𠜎"
[8] pry(main)> mixed_utf_8 == orig_mixed_utf_8
=> false
```

`Etc` is a special case for us.  Consider the same scenario above - ie host in ISO-8859-1 managed by Puppet, but instead believing the non-UTF-8 encoding is inaccurate and doing `force_encoding!`. 

Puppet will be managing the properties of a user, but Puppet has not run yet on the host. The values on disk or accurate an accurate representation of their encoding, ie ISO-8859-1. When Puppet first runs, these values are read in and incorrectly re-labeled (overridden) as UTF-8 by Puppet::Etc. Puppet compares these values to the the UTF-8 `should` values, finds them out of sync, and overwrites them on the system with the UTF-8 values. The next time Puppet runs, the UTF-8 values are read in as ISO-8859-1, but correctly overridden by Puppet::Etc to UTF-8, and Puppet correctly finds them in sync with the catalog `should` values. Even if we had not incorrectly overridden the encoding on the very first Puppet run, the result would have been the same - Puppet would have found the values out of sync, as the ISO-8859-1 value would not equal the UTF-8 value, and corrected them.